### PR TITLE
Complete shopping cart stories.

### PR DESCRIPTION
**Pull request by Manpreet-Singh001**
* Generates shopping cart based on your storage and mealplan while taking expiry date in account.
* Fixed the duplicate item generated issue. 
* User can add partial items picked up to their storage.
* Resilient to user entering wrong fields in the item they want to send to storage.
* Smart Unit conversions.
* Pick up system recognizes if user picked up partial items and allows them to do so.

### CODE REVIEW

File: app/src/main/java/com/xffffff/wellfed/common/DateUtil.java [hunk 1]
```diff
@@ -145,4 +145,15 @@ public String format(Date date, String pattern) {
...
         return simpleDateFormat.format(date);
     }
+
+    public int format(Date date){
+        String formatted = hashFormat.format(date);
+        String[] helper =  formatted.split("-");
+        StringBuilder needed = new StringBuilder();
+        for (String s: helper){
+            needed.append(s);
+        }
```
Comment (steventango) [inline app/src/main/java/com/xffffff/wellfed/common/DateUtil.java:155]: can replace with `needed = format(date, "yyyyMMdd")`

File: app/src/main/java/com/xffffff/wellfed/shoppingcart/ShoppingCartIngredientAdapter.java [hunk 2]
```diff
@@ -52,14 +61,41 @@ public ShoppingCartIngredientAdapter(ShoppingCartDB db) {
...
+        String id = doc.getId();
+        boolean isPickedUp = (Boolean) doc.getData().get("picked");
+        Double amount = ((Double) doc.getData().get("amount"));
+
+        ShoppingCartIngredient shoppingCartIngredient = new ShoppingCartIngredient(description);
+        shoppingCartIngredient.setId(id);
+        shoppingCartIngredient.setCategory(category);
+        shoppingCartIngredient.setUnit(unit);
+        shoppingCartIngredient.setPickedUp(isPickedUp);
+        shoppingCartIngredient.setAmount(amount);
```
Comment (steventango) [inline app/src/main/java/com/xffffff/wellfed/shoppingcart/ShoppingCartIngredientAdapter.java:80]: move this to DB

File: app/src/main/java/com/xffffff/wellfed/shoppingcart/ShoppingCartIngredientAdapter.java [hunk 3]
```diff
@@ -52,14 +61,41 @@ public ShoppingCartIngredientAdapter(ShoppingCartDB db) {
...
+
+        holder.description.setText(description);
+        String amountText = String.format(Locale.CANADA, "%.2f", amount);
+        holder.subtext.setText(amountText + " " + unit + " | " + category);
+        holder.checkBox.setChecked(isPickedUp);
+        if (holder.checkBox.isChecked()) {
+
+        }
```
Comment (steventango) [inline app/src/main/java/com/xffffff/wellfed/shoppingcart/ShoppingCartIngredientAdapter.java:88]: what is this if statement for

File: app/src/main/java/com/xffffff/wellfed/shoppingcart/ShoppingCartIngredientController.java [hunk 4]
```diff
@@ -230,12 +253,125 @@ public void updateIngredientInShoppingCart(
...
+            Pair<Double, String> val =
+                    bestUnitHelper(ingredient, unitHelper);
+            if (required.get(val.second) == null) {
+                ArrayList<Pair<Integer, Double>> listNeeded = new ArrayList<>();
+                listNeeded
+                        .add(new Pair<Integer, Double>(dateUtil.format(mealPlan.getEatDate()), val.first));
+                required.put(val.second, listNeeded);
+            } else {
+                required.get(val.second)
+                        .add(new Pair<Integer, Double>(dateUtil.format(mealPlan.getEatDate()), val.first));
```
Comment (steventango) [inline app/src/main/java/com/xffffff/wellfed/shoppingcart/ShoppingCartIngredientController.java:339]: can you combine this logic with the generate required from ingredients?

**Review by steventango — APPROVED** [review]

**Comment by Manpreet-Singh001** [comment]
fixes: #211

**Comment by Manpreet-Singh001** [comment]
fixes #17

**Comment by Manpreet-Singh001** [comment]
fixes: #18 #19 #20 #21

**Pull request merged**